### PR TITLE
Anko - Update logic register

### PR DESCRIPTION
**Pull request by at-haingo**
#### Related issues (Required)
- Task #

#### Description (Required)
- [x] Update logic register

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 1]
```diff
@@ -26,192 +25,147 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+                                .subscribe(this::handleUpdateUserCompleted))
+                    }
                 }
-            }.show()
+            }
         }
     }
 
-    private fun onGetUser(user: User) {
+    private fun handleGetUserCompleted(user: User) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: WARNING: Excessive nesting leads to hidden complexity. Prefer extracting code to make it easier to understand.

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 2]
```diff
@@ -255,50 +208,58 @@ class RegisterActivity : BaseActivity() {
...
-    private fun invisibleProgressBar(progressBar: ProgressBar) {
-        progressBar.visibility = View.GONE
-        window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
+    private fun handleSkipClicked(userParam: UserParams) {
+        alert {
+            title = getString(R.string.register_title_dialog)
+            message = getString(R.string.register_message_dialog)
+            yesButton {
+                addDisposables(registerViewModel.createUser(userParam)
+                        .subscribe(this@RegisterActivity::handleCreateUserCompleted))
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: WARNING: Expression with labels increase complexity and affect maintainability.

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 3]
```diff
@@ -26,192 +25,147 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
-        val disposable = registerViewModel.createUser(userParam)
-                .subscribe(this::onUserCreated)
-        addDisposables(disposable)
+                            override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
+                                avatarBitmap = bitmap
+                                handleGetAvatarCompleted()
+                            }
+                        })
+            } else {
+                avatarBitmap = data.extras.get("data") as Bitmap
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: ERROR: Cast operator throws an exception if the cast is not possible.

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 4]
```diff
@@ -26,192 +25,147 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
         super.onActivityResult(requestCode, resultCode, data)
         if (resultCode == Activity.RESULT_OK && requestCode == REQUEST_CODE_PICK_IMAGE && data != null) {
-            setUserAvatar(data)
-        }
-    }
+            val uri = data.data
+            if (uri != null) {
+                Picasso.with(this)
+                        .load(uri)
+                        .resize(300, 300)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: WARNING: Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers.

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 5]
```diff
@@ -26,192 +25,148 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
-        } else {
-            if (ui.edtName.text.toString() != userWay?.name
-                    || ui.edtPhone.text.toString() != userWay?.phone?.removeRange(0, NUM_CHAR_REMOVE)
-                    || avatarUri != null) {
-                updateUser(userParam)
-            }
-        }
-        startActivity<HomeActivity>()
-    }
+    internal fun eventClicked(view: View) {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: change method name to `eventOnViewClicked`

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 6]
```diff
@@ -26,192 +25,148 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
         private const val AVATAR_SIZE = 300
-        private const val RESULT_SUCCESS = "Success"
+        private const val KEY_FROM_REGISTER = "Register"
     }
 
-    private var countries: MutableList<Country> = ArrayList()
-    internal lateinit var registerViewModel: RegisterViewModel
     private lateinit var ui: RegisterActivityUI
+    private lateinit var registerViewModel: RegisterViewModel
+    private val countries: MutableList<Country> = ArrayList()
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: Please use `mutableListOf`

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 7]
```diff
@@ -26,192 +25,148 @@ import vn.asiantech.way.data.source.remote.response.ResponseStatus
...
+        registerViewModel.eventBackPressed()
     }
 
     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         super.onActivityResult(requestCode, resultCode, data)
         if (resultCode == Activity.RESULT_OK && requestCode == REQUEST_CODE_PICK_IMAGE && data != null) {
-            setUserAvatar(data)
-        }
-    }
+            val uri = data.data
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: Please handle `onPermissionResult`

**Comment by at-vinhhuynh** [comment]
@at-tienhoang 
I finish review this PR, please check it again.
thank you!

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 8]
```diff
@@ -105,113 +92,89 @@ class RegisterActivity : BaseActivity() {
...
-            startActivity<HomeActivity>()
-            toast(responseStatus.message)
-        } else {
-            alert {
-                title = getString(R.string.dialog_title_error)
-                message = updateMessage
-                yesButton {
-                    title = getString(R.string.dialog_button_ok)
-                    it.dismiss()
+                if (registerViewModel.isRegister) {
```
Comment (at-tienhoang) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: Move this block into viewmodel

File: app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt [hunk 9]
```diff
@@ -105,114 +91,79 @@ class RegisterActivity : BaseActivity() {
...
+                            .subscribe(this::handleUpdateUserCompleted))
                 }
-            }.show()
+            }
         }
     }
 
-    private fun onGetUser(user: User) {
-        userWay = user
+    private fun handleGetUserCompleted(user: User) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/ui/register/RegisterActivity.kt]: WARNING: Excessive nesting leads to hidden complexity. Prefer extracting code to make it easier to understand.

File: app/src/main/java/vn/asiantech/way/extension/EditText.kt [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import android.text.Editable
+import android.text.TextWatcher
+import android.widget.EditText
+
+/**
+ *
+ * Created by haingoq on 06/12/2017.
+ */
+fun EditText.onTextChangeListener(onTextChanged: (CharSequence?) -> Unit = {},
+                                     afterTextChanged: (Editable?) -> Unit = {},
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/extension/EditText.kt]: WARNING: Unexpected indentation

File: app/src/main/java/vn/asiantech/way/extension/EditText.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+                                     beforeTextChanged: (CharSequence?) -> Unit = {}) {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/extension/EditText.kt]: WARNING: Unexpected indentation

**Pull request merged**